golem/event_bus/in_memory/event_bus.py

```diff
@@ -0,0 +1,226 @@
+import asyncio
+import logging
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Awaitable, Callable, DefaultDict, List, Optional, Tuple, Type
+
+from golem.event_bus.base import Event, EventBus, EventBusError, TEvent
+from golem.utils.asyncio import create_task_with_logging
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class _CallbackInfo:
+    callback: Callable[[TEvent], Awaitable[None]]
+    filter_func: Optional[Callable[[TEvent], bool]]
+    once: bool
+
+
+_CallbackHandler = Tuple[Type[TEvent], _CallbackInfo]
+
+
+class InMemoryEventBus(EventBus[_CallbackHandler]):
+    def __init__(self):
+        self._callbacks: DefaultDict[Type[Event], List[_CallbackInfo]] = defaultdict(list)
+        self._event_queue: asyncio.Queue[Event] = asyncio.Queue()
+        self._process_event_queue_loop_task: Optional[asyncio.Task] = None
+
+    async def start(self):
+        logger.debug("Starting event bus...")
+
+        if self.is_started():
+            message = "Event bus is already started!"
+            logger.debug(f"Starting event bus failed with `{message}`")
+            raise EventBusError(message)
+
+        self._process_event_queue_loop_task = create_task_with_logging(
+            self._process_event_queue_loop()
+        )
+
+        logger.debug("Starting event bus done")
+
+    async def stop(self):
+        logger.debug("Stopping event bus...")
+
+        if not self.is_started():
+            message = "Event bus is not started!"
+            logger.debug(f"Stopping event bus failed with `{message}`")
+            raise EventBusError(message)
+
+        await self._event_queue.join()
+
+        if self._process_event_queue_loop_task is not None:
+            self._process_event_queue_loop_task.cancel()
+            self._process_event_queue_loop_task = None
+
+        logger.debug("Stopping event bus done")
+
+    def is_started(self) -> bool:
+        return (
+            self._process_event_queue_loop_task is not None
+            and not self._process_event_queue_loop_task.done()
+        )
+
+    async def on(
+        self,
+        event_type: Type[TEvent],
+        callback: Callable[[TEvent], Awaitable[None]],
+        filter_func: Optional[Callable[[TEvent], bool]] = None,
+    ) -> _CallbackHandler:
+        logger.debug(
+            f"Adding callback handler for `{event_type}` with callback `{callback}`"
+            f" and filter `{filter_func}`..."
+        )
+
+        callback_info = _CallbackInfo(
+            callback=callback,
+            filter_func=filter_func,
+            once=False,
+        )
+
+        self._callbacks[event_type].append(callback_info)
+
+        callback_handler = (event_type, callback_info)
+
+        logger.debug(f"Adding callback handler done with `{id(callback_handler)}`")
```

Comment:
I'd merge this debug message with the previous one (on line 72) so that it's obvious from that single line what handler is being registered and what id it gets... otherwise, this line here carries too little information on its own...